light/beacon/binary_merkle_test.go

```diff
@@ -0,0 +1,264 @@
+		}
+		if !ok {
+			t.Errorf("VerifySingleProof length invalid (index = %d)", index)
+		}
+	}
+}
+
+func TestMultiProof(t *testing.T) {
+	for count := 0; count < 300; count++ {
+		failIndex := uint64(128 + rand.Intn(128))
+		indexList := make([]uint64, 10)
+		for i := range indexList {
+			for {
+				indexList[i] = uint64(128 + rand.Intn(128))
+				if indexList[i]^failIndex > 1 {
+					// failIndex should not be available in proofs so it should not be equal to or sibling of a stored index
+					break
+				}
+			}
+		}
+
+		readers := make([]ProofReader, len(indexList))
+		for i, index := range indexList {
+			var mp MultiProof
+			mp.Format = NewIndexMapFormat().AddLeaf(index, nil)
+			writer := NewMultiProofWriter(mp.Format, &mp.Values, nil)
+			testTraverseProof(t, testProofReader, writer, true)
+			readers[i] = mp.Reader(nil)
+		}
+
+		// create a single multiproof from the merged reader using a subset of indices
+		var mp MultiProof
+		format := NewIndexMapFormat()
+		mpCount := rand.Intn(11) // add a subset of indices to the created multiproof format
+		for i := 0; i < mpCount; i++ {
+			format.AddLeaf(indexList[i], nil)
+		}
+		mp.Format = format
+		expSuccess := rand.Intn(2) == 0
+		if !expSuccess {
+			// add an index that should not be available in the merged reader, expect the traversal to fail
+			format.AddLeaf(failIndex, nil)
+		}
+		testTraverseProof(t, MergedReader(readers), NewMultiProofWriter(format, &mp.Values, nil), expSuccess)
+
+		if expSuccess {
+			mpwCount := rand.Intn(mpCount + 1) // create writers for a subset of the previously selected indices (available in mp)
+			mps := make([]MultiProof, mpwCount)
+			writers := make([]ProofWriter, mpwCount)
+			for i := range mps {
+				mps[i].Format = NewIndexMapFormat().AddLeaf(indexList[i], nil)
+				writers[i] = NewMultiProofWriter(mps[i].Format, &mps[i].Values, nil)
+			}
+			reader := mp.Reader(nil)
+			testTraverseProof(t, reader, MergedWriter(writers), true)
+			if !reader.Finished() {
+				t.Errorf("MultiProofReader not finished")
+			}
+			// test individual single-value multiproofs
+			for i, mp := range mps {
+				if valueIndex, ok := ProofFormatIndexMap(mp.Format)[indexList[i]]; !ok || mp.Values[valueIndex] != testMerkleTree[indexList[i]] {
+					t.Errorf("Could not find tree index %d in single-value multiproof", indexList[i])
+				}
+			}
+		}
+	}
+}
+
+func testTraverseProof(t *testing.T, reader ProofReader, writer ProofWriter, expSuccess bool) {
+	root, ok := TraverseProof(reader, writer)
+	if expSuccess {
+		if root != common.Hash(testMerkleTree[1]) {
+			t.Errorf("TraverseProof root hash mismatch")
+		}
+		if !ok {
+			t.Errorf("TraverseProof insufficient reader data")
+		}
+	} else if ok {
+		t.Errorf("TraverseProof succeeded (expected to fail)")
+	}
+}
+
+func formatsEqual(f1, f2 ProofFormat) bool {
+	if f1 == nil && f2 == nil {
+		return true
+	}
+	if f1 == nil || f2 == nil {
+		return true
```

Comment:
return false

Reply:
Indeed, and also facepalm :) Turns out one of the tests was actually wrong, failed after fixing this return value. It's also fixed now.